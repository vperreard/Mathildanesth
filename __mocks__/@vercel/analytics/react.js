// Mock pour @vercel/analytics/react
const React = require('react');

const Analytics = ({ children, ...props }) => {
  return React.createElement('div', { 'data-testid': 'analytics-mock', ...props }, children);
};

const track = jest.fn();

module.exports = {
  __esModule: true,
  Analytics,
  track,
  default: Analytics,
};