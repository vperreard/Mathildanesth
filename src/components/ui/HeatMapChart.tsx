'use client';

import React, { useState, useEffect, useMemo } from 'react';
import { Card, CardContent, CardHeader, CardTitle, CardDescription } from '@/components/ui/card';
import { Select, SelectContent, SelectItem, SelectTrigger, SelectValue } from '@/components/ui/select';
import { Skeleton } from '@/components/ui/skeleton';
import { HelpCircle } from 'lucide-react';
import { Tooltip, TooltipContent, TooltipProvider, TooltipTrigger } from '@/components/ui/tooltip';

interface CellData {
    x: string;
    y: string;
    value: number;
    label?: string;
    description?: string;
}

interface HeatMapProps {
    title?: string;
    description?: string;
    data: CellData[];
    xAxisLabel?: string;
    yAxisLabel?: string;
    colorRanges?: ColorRange[];
    isLoading?: boolean;
    width?: number | string;
    height?: number | string;
    showLegend?: boolean;
    onCellClick?: (cell: CellData) => void;
}

interface ColorRange {
    min: number;
    max: number;
    color: string;
}

// Palette de couleurs par défaut
const DEFAULT_COLOR_RANGES: ColorRange[] = [
    { min: 0, max: 20, color: '#f3f4f6' },      // Gris très clair
    { min: 20, max: 40, color: '#dcfce7' },     // Vert très clair
    { min: 40, max: 60, color: '#86efac' },     // Vert clair
    { min: 60, max: 80, color: '#22c55e' },     // Vert
    { min: 80, max: 90, color: '#15803d' },     // Vert foncé
    { min: 90, max: 100, color: '#166534' },    // Vert très foncé
];

// Palette pour les surcharges
const OVERLOAD_COLOR_RANGES: ColorRange[] = [
    { min: 0, max: 60, color: '#f3f4f6' },      // Gris très clair
    { min: 60, max: 80, color: '#fef9c3' },     // Jaune très clair
    { min: 80, max: 90, color: '#fde047' },     // Jaune
    { min: 90, max: 100, color: '#facc15' },    // Jaune foncé
    { min: 100, max: 110, color: '#f97316' },   // Orange
    { min: 110, max: 500, color: '#dc2626' },   // Rouge
];

const PALETTE_OPTIONS = [
    { id: 'default', name: 'Standard', ranges: DEFAULT_COLOR_RANGES },
    { id: 'overload', name: 'Surcharge', ranges: OVERLOAD_COLOR_RANGES },
];

const HeatMapChart = ({
    title = 'Carte de chaleur',
    description = 'Visualisation de la charge de travail',
    data,
    xAxisLabel = 'Période',
    yAxisLabel = 'Ressource',
    colorRanges = DEFAULT_COLOR_RANGES,
    isLoading = false,
    width = '100%',
    height = 400,
    showLegend = true,
    onCellClick
}: HeatMapProps) => {
    const [hoverCell, setHoverCell] = useState<CellData | null>(null);
    const [selectedPalette, setSelectedPalette] = useState<string>('default');

    // Extraire les valeurs uniques des axes X et Y
    const xValues = useMemo(() => {
        return [...new Set(data.map(item => item.x))].sort();
    }, [data]);

    const yValues = useMemo(() => {
        return [...new Set(data.map(item => item.y))].sort();
    }, [data]);

    // Déterminer la couleur de la cellule en fonction de sa valeur
    const getCellColor = (value: number, ranges: ColorRange[]) => {
        const range = ranges.find(r => value >= r.min && value < r.max);
        return range ? range.color : ranges[ranges.length - 1].color;
    };

    // Obtenir le texte de la cellule
    const getCellText = (value: number) => {
        if (value === 0) return '-';
        return `${value}%`;
    };

    // Calculer la largeur et la hauteur des cellules
    const cellWidth = `calc(${100 / (xValues.length || 1)}%)`;
    const cellHeight = 40;  // hauteur fixe, en pixels

    // Utilisez la palette sélectionnée
    const activePalette = useMemo(() => {
        const palette = PALETTE_OPTIONS.find(p => p.id === selectedPalette);
        return palette ? palette.ranges : DEFAULT_COLOR_RANGES;
    }, [selectedPalette]);

    // Préparation des données pour l'affichage sous forme de grille
    const gridData = useMemo(() => {
        const grid: (CellData | null)[][] = [];

        // Initialiser la grille avec des valeurs nulles
        for (let i = 0; i < yValues.length; i++) {
            grid[i] = [];
            for (let j = 0; j < xValues.length; j++) {
                grid[i][j] = null;
            }
        }

        // Remplir la grille avec les données
        data.forEach(item => {
            const rowIndex = yValues.indexOf(item.y);
            const colIndex = xValues.indexOf(item.x);
            if (rowIndex >= 0 && colIndex >= 0) {
                grid[rowIndex][colIndex] = item;
            }
        });

        return grid;
    }, [data, xValues, yValues]);

    if (isLoading) {
        return (
            <Card>
                <CardHeader>
                    <CardTitle><Skeleton className="h-8 w-1/3" /></CardTitle>
                    <CardDescription><Skeleton className="h-4 w-2/3" /></CardDescription>
                </CardHeader>
                <CardContent>
                    <div style={{ height: height, width: width }} className="relative">
                        <Skeleton className="h-full w-full" />
                    </div>
                </CardContent>
            </Card>
        );
    }

    return (
        <Card className="overflow-hidden">
            <CardHeader className="pb-2">
                <div className="flex justify-between items-start">
                    <div>
                        <CardTitle>{title}</CardTitle>
                        <CardDescription>{description}</CardDescription>
                    </div>

                    <div className="flex items-center gap-2">
                        <Select value={selectedPalette} onValueChange={setSelectedPalette}>
                            <SelectTrigger className="w-[140px]">
                                <SelectValue placeholder="Palette de couleurs" />
                            </SelectTrigger>
                            <SelectContent>
                                {PALETTE_OPTIONS.map((palette) => (
                                    <SelectItem key={palette.id} value={palette.id}>
                                        {palette.name}
                                    </SelectItem>
                                ))}
                            </SelectContent>
                        </Select>

                        <TooltipProvider>
                            <Tooltip>
                                <TooltipTrigger asChild>
                                    <HelpCircle size={18} className="text-muted-foreground cursor-help" />
                                </TooltipTrigger>
                                <TooltipContent className="max-w-xs">
                                    <p>Palette "Standard" : Mesure la couverture à partir de 0% (vide) à 100% (couverture complète).</p>
                                    <p className="mt-1">Palette "Surcharge" : Inclut les valeurs au-delà de 100% pour identifier les surcharges.</p>
                                </TooltipContent>
                            </Tooltip>
                        </TooltipProvider>
                    </div>
                </div>
            </CardHeader>
            <CardContent>
                <div className="flex flex-col" style={{ height, width, overflow: 'auto' }}>
                    {/* En-têtes des colonnes (axe X) */}
                    <div className="flex pl-24">
                        {xValues.map((value, index) => (
                            <div
                                key={`header-${index}`}
                                className="flex-shrink-0 text-center text-xs font-medium text-gray-600 pb-1"
                                style={{ width: cellWidth }}
                            >
                                {value}
                            </div>
                        ))}
                    </div>

                    {/* Grille principale avec les noms des lignes (axe Y) */}
                    <div className="flex-grow overflow-auto">
                        {gridData.map((row, rowIndex) => (
                            <div key={`row-${rowIndex}`} className="flex items-center h-10 mb-px">
                                <div className="flex-shrink-0 w-24 pr-2 text-xs font-medium text-gray-600 text-right whitespace-nowrap overflow-hidden text-ellipsis">
                                    {yValues[rowIndex]}
                                </div>
                                <div className="flex flex-grow">
                                    {row.map((cell, colIndex) => {
                                        // Valeur par défaut si la cellule est null
                                        const value = cell ? cell.value : 0;
                                        const cellColor = getCellColor(value, activePalette);
                                        const cellText = getCellText(value);

                                        return (
                                            <div
                                                key={`cell-${rowIndex}-${colIndex}`}
                                                className="flex-shrink-0 flex items-center justify-center border border-gray-200 text-xs font-medium relative cursor-pointer transition-all duration-100"
                                                style={{
                                                    width: cellWidth,
                                                    height: `${cellHeight}px`,
                                                    backgroundColor: cellColor,
                                                }}
                                                onMouseEnter={() => cell && setHoverCell(cell)}
                                                onMouseLeave={() => setHoverCell(null)}
                                                onClick={() => cell && onCellClick && onCellClick(cell)}
                                            >
                                                {cellText}

                                                {/* Tooltip informatif au survol */}
                                                {hoverCell === cell && cell && (
                                                    <div className="absolute z-10 bottom-full left-1/2 transform -translate-x-1/2 mb-2 bg-gray-800 text-white text-xs rounded py-1 px-2 shadow-lg whitespace-nowrap">
                                                        <div><strong>{cell.x} - {cell.y}</strong></div>
                                                        <div>Valeur: {cell.value}%</div>
                                                        {cell.description && <div>{cell.description}</div>}
                                                    </div>
                                                )}
                                            </div>
                                        );
                                    })}
                                </div>
                            </div>
                        ))}
                    </div>
                </div>

                {showLegend && (
                    <div className="mt-4 flex items-center">
                        <span className="text-xs text-gray-600 mr-2">Légende:</span>
                        <div className="flex items-center">
                            {activePalette.map((range, index) => (
                                <div key={`legend-${index}`} className="flex items-center mx-1">
                                    <div
                                        className="w-4 h-4 mr-1 border border-gray-300"
                                        style={{ backgroundColor: range.color }}
                                    ></div>
                                    <span className="text-xs text-gray-600">{range.min}-{range.max}%</span>
                                </div>
                            ))}
                        </div>
                    </div>
                )}
            </CardContent>
        </Card>
    );
};

export default HeatMapChart; 