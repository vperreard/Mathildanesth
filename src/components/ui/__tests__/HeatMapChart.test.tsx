import React from 'react';
import { render, screen } from '@testing-library/react';
import HeatMapChart from '../HeatMapChart';

// Mock des composants UI
jest.mock('@/components/ui/card', () => ({
    Card: ({ children, ...props }: any) => <div {...props}>{children}</div>,
    CardContent: ({ children, ...props }: any) => <div {...props}>{children}</div>,
    CardHeader: ({ children, ...props }: any) => <div {...props}>{children}</div>,
    CardTitle: ({ children, ...props }: any) => <h3 {...props}>{children}</h3>,
    CardDescription: ({ children, ...props }: any) => <p {...props}>{children}</p>,
}));

jest.mock('@/components/ui/select', () => ({
    Select: ({ children, ...props }: any) => <div {...props}>{children}</div>,
    SelectContent: ({ children, ...props }: any) => <div {...props}>{children}</div>,
    SelectItem: ({ children, ...props }: any) => <div {...props}>{children}</div>,
    SelectTrigger: ({ children, ...props }: any) => <button {...props}>{children}</button>,
    SelectValue: ({ ...props }: any) => <span {...props} />,
}));

jest.mock('@/components/ui/skeleton', () => ({
    Skeleton: ({ className }: any) => <div className={className} data-testid="skeleton" />,
}));

jest.mock('@/components/ui/tooltip', () => ({
    Tooltip: ({ children }: any) => <div>{children}</div>,
    TooltipContent: ({ children }: any) => <div>{children}</div>,
    TooltipProvider: ({ children }: any) => <div>{children}</div>,
    TooltipTrigger: ({ children }: any) => <div>{children}</div>,
}));

describe('HeatMapChart', () => {
    // Données de test
    const mockData = [
        { x: 'A', y: '1', value: 100 },
        { x: 'B', y: '1', value: 80 },
        { x: 'A', y: '2', value: 60 },
        { x: 'B', y: '2', value: 40 },
    ];

    const defaultProps = {
        data: mockData,
        xAxisLabel: 'X Axis',
        yAxisLabel: 'Y Axis',
        title: 'Test Heat Map',
        description: 'Test Description',
    };

    it('doit rendre correctement avec les props par défaut', () => {
        // Act
        render(<HeatMapChart {...defaultProps} />);

        // Assert
        expect(screen.getByText('Test Heat Map')).toBeInTheDocument();
        expect(screen.getByText('Test Description')).toBeInTheDocument();
        // Le composant custom n'utilise pas responsive-container
        // Vérifier que le composant se rend sans erreur
        expect(screen.getByText('Test Heat Map')).toBeTruthy();
    });

    it('doit rendre le conteneur sans crasher quand data est vide', () => {
        // Act
        render(<HeatMapChart {...defaultProps} data={[]} />);

        // Assert
        expect(screen.getByText('Test Heat Map')).toBeInTheDocument();
        expect(screen.getByText('Aucune donnée disponible')).toBeInTheDocument();
    });

    it('doit appliquer la bonne échelle de couleurs', () => {
        // Act
        const { container } = render(
            <HeatMapChart
                {...defaultProps}
                colorScale={['#ff0000', '#00ff00']}
            />
        );

        // Vérifier que l'échelle de couleurs est bien définie
        // Note: C'est difficile de tester cela précisément sans mocks plus profonds de recharts
        // ou sans une approche de snapshot testing
        expect(container).toMatchSnapshot();
    });

    it('doit utiliser des étiquettes d\'axe personnalisées', () => {
        // Act
        render(
            <HeatMapChart
                {...defaultProps}
                xAxisLabel="Custom X"
                yAxisLabel="Custom Y"
            />
        );

        // Assert
        expect(screen.getByText('Custom X')).toBeInTheDocument();
        expect(screen.getByText('Custom Y')).toBeInTheDocument();
    });

    it('doit afficher un titre personnalisé', () => {
        // Act
        render(
            <HeatMapChart
                {...defaultProps}
                title="Titre Personnalisé"
            />
        );

        // Assert
        expect(screen.getByText('Titre Personnalisé')).toBeInTheDocument();
    });

    it('doit utiliser une taille personnalisée si fournie', () => {
        // Act
        const { container } = render(
            <HeatMapChart
                {...defaultProps}
                width={800}
                height={600}
            />
        );

        // Assert - Vérifier que le style est appliqué
        // Pour un test plus précis, on pourrait vérifier les styles appliqués
        expect(container.firstChild).toHaveStyle('width: 800px; height: 600px;');
    });
}); 