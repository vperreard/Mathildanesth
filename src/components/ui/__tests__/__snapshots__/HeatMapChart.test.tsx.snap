// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeatMapChart doit appliquer la bonne échelle de couleurs 1`] = `
<div>
  <div
    class="overflow-hidden"
  >
    <div
      class="pb-2"
    >
      <div
        class="flex justify-between items-start"
      >
        <div>
          <h3>
            Test Heat Map
          </h3>
          <p>
            Test Description
          </p>
        </div>
        <div
          class="flex items-center gap-2"
        >
          <div
            value="default"
          >
            <button
              class="w-[140px]"
            >
              <span
                placeholder="Palette de couleurs"
              />
            </button>
            <div>
              <div
                value="default"
              >
                Standard
              </div>
              <div
                value="overload"
              >
                Surcharge
              </div>
            </div>
          </div>
          <div>
            <div>
              <div>
                <svg
                  aria-hidden="true"
                  class="lucide lucide-circle-help text-muted-foreground cursor-help"
                  fill="none"
                  height="18"
                  stroke="currentColor"
                  stroke-linecap="round"
                  stroke-linejoin="round"
                  stroke-width="2"
                  viewBox="0 0 24 24"
                  width="18"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <circle
                    cx="12"
                    cy="12"
                    r="10"
                  />
                  <path
                    d="M9.09 9a3 3 0 0 1 5.83 1c0 2-3 3-3 3"
                  />
                  <path
                    d="M12 17h.01"
                  />
                </svg>
              </div>
              <div>
                <p>
                  Palette "Standard" : Mesure la couverture à partir de 0% (vide) à 100% (couverture complète).
                </p>
                <p
                  class="mt-1"
                >
                  Palette "Surcharge" : Inclut les valeurs au-delà de 100% pour identifier les surcharges.
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div>
      <div
        class="flex flex-col"
        style="height: 400px; width: 100%; overflow: auto;"
      >
        <div
          class="flex pl-24"
        >
          <div
            class="flex-shrink-0 text-center text-xs font-medium text-gray-600 pb-1"
            style="width: calc(50%);"
          >
            A
          </div>
          <div
            class="flex-shrink-0 text-center text-xs font-medium text-gray-600 pb-1"
            style="width: calc(50%);"
          >
            B
          </div>
        </div>
        <div
          class="flex-grow overflow-auto"
        >
          <div
            class="flex items-center h-10 mb-px"
          >
            <div
              class="flex-shrink-0 w-24 pr-2 text-xs font-medium text-gray-600 text-right whitespace-nowrap overflow-hidden text-ellipsis"
            >
              1
            </div>
            <div
              class="flex flex-grow"
            >
              <div
                class="flex-shrink-0 flex items-center justify-center border border-gray-200 text-xs font-medium relative cursor-pointer transition-all duration-100"
                style="width: calc(50%); height: 40px; background-color: rgb(22, 101, 52);"
              >
                100%
              </div>
              <div
                class="flex-shrink-0 flex items-center justify-center border border-gray-200 text-xs font-medium relative cursor-pointer transition-all duration-100"
                style="width: calc(50%); height: 40px; background-color: rgb(21, 128, 61);"
              >
                80%
              </div>
            </div>
          </div>
          <div
            class="flex items-center h-10 mb-px"
          >
            <div
              class="flex-shrink-0 w-24 pr-2 text-xs font-medium text-gray-600 text-right whitespace-nowrap overflow-hidden text-ellipsis"
            >
              2
            </div>
            <div
              class="flex flex-grow"
            >
              <div
                class="flex-shrink-0 flex items-center justify-center border border-gray-200 text-xs font-medium relative cursor-pointer transition-all duration-100"
                style="width: calc(50%); height: 40px; background-color: rgb(34, 197, 94);"
              >
                60%
              </div>
              <div
                class="flex-shrink-0 flex items-center justify-center border border-gray-200 text-xs font-medium relative cursor-pointer transition-all duration-100"
                style="width: calc(50%); height: 40px; background-color: rgb(134, 239, 172);"
              >
                40%
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="mt-4 flex items-center"
      >
        <span
          class="text-xs text-gray-600 mr-2"
        >
          Légende:
        </span>
        <div
          class="flex items-center"
        >
          <div
            class="flex items-center mx-1"
          >
            <div
              class="w-4 h-4 mr-1 border border-gray-300"
              style="background-color: rgb(243, 244, 246);"
            />
            <span
              class="text-xs text-gray-600"
            >
              0
              -
              20
              %
            </span>
          </div>
          <div
            class="flex items-center mx-1"
          >
            <div
              class="w-4 h-4 mr-1 border border-gray-300"
              style="background-color: rgb(220, 252, 231);"
            />
            <span
              class="text-xs text-gray-600"
            >
              20
              -
              40
              %
            </span>
          </div>
          <div
            class="flex items-center mx-1"
          >
            <div
              class="w-4 h-4 mr-1 border border-gray-300"
              style="background-color: rgb(134, 239, 172);"
            />
            <span
              class="text-xs text-gray-600"
            >
              40
              -
              60
              %
            </span>
          </div>
          <div
            class="flex items-center mx-1"
          >
            <div
              class="w-4 h-4 mr-1 border border-gray-300"
              style="background-color: rgb(34, 197, 94);"
            />
            <span
              class="text-xs text-gray-600"
            >
              60
              -
              80
              %
            </span>
          </div>
          <div
            class="flex items-center mx-1"
          >
            <div
              class="w-4 h-4 mr-1 border border-gray-300"
              style="background-color: rgb(21, 128, 61);"
            />
            <span
              class="text-xs text-gray-600"
            >
              80
              -
              90
              %
            </span>
          </div>
          <div
            class="flex items-center mx-1"
          >
            <div
              class="w-4 h-4 mr-1 border border-gray-300"
              style="background-color: rgb(22, 101, 52);"
            />
            <span
              class="text-xs text-gray-600"
            >
              90
              -
              100
              %
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
